Fixed #11154, #22270 -- Made proxy model permissions use correct content type.

arthurio opened the pull request:
# Context

Previous PR: #4681

**Ticket links**:
- Main bug: https://code.djangoproject.com/ticket/11154
- Maybe related bug: https://code.djangoproject.com/ticket/17904
- Docs: https://code.djangoproject.com/ticket/22270

# State of things before this patch

## Permissions for proxy models

They are created using the concrete model content type.

```python
# myapp/model.py
class User(models.Model):
    pass

class Student(models.Model):
    class Meta:
        proxy = True
        # Extra permissions
        permissions = (
            ('email_student', 'Can email students'),
        )

student_permission = Permission.objects.get(codename='email_student')
user_permission = Permission.objects.get(codename='add_user')

# The content type used for the permissions is the one of the concrete model
assert student_permission.content_type == user_permission.content_type

# Add the permissions to the user
user.user_permissions.add(student_permission)
user.user_permissions.add(user_permission)

# Then check if it worked as you would expect
assert user.has_perms('myapp.email_student', 'myapp.add_user') is True
```

So far everything works as expected... But now let's use a proxy model in a different app!

```python
# myotherapp/models.py
from myapp.models import User

class Teacher(User):
    class Meta:
        proxy = True

teacher_permission = Permission.objects.get(codename='add_teacher')
assert teacher_permission.content_type == user_permission.content_type

# Add the permission to the user
user.user_permissions.add(teacher_permission)

# And here is the catch...
assert user.has_perm('myotherapp.add_teacher') is True, 'Bad luck, you have to use the app_label of the concrete model'
# Instead
assert user.has_perm('myapp.add_teacher') is True
```
## Admin pages

For proxy models having a different `app_label` than the concrete model would, their admin page would not be accessible to users having the proxy model permissions because of how the permission lookup is done in `django.contrib.admin.options.py`. As mentioned in the previous example, you need to use the `app_label` of the concrete model to lookup the permission, but the admin was constructing the string with the model app_label whether or not that model was a proxy model.
```python
# myapp/admin.py
from myapp.models import Student

admin.site.register(Student)
```

```python
# myotherapp/admin.py
from myotherapp.models import Teacher


admin.site.register(Teacher)
```

```python
user.user_permissions.add(Permission.objects.get(codename='change_student'))
user.user_permissions.add(Permission.objects.get(codename='change_teacher'))
# Log the user in...
# ...
response = client.get('admin:myapp_student_changelist')
assert response.status_code == 200

response = client.get('admin:myotherapp_teacher_changelist')
assert response.status_code == 200, 'Nope, it returns 403 because of the bad permission lookup!'
```

## Existing workarounds (That I know of...)
So currently, to give users access to the `Teacher` admin page, there are 2 options:
- Create manually the permissions that will use the proxy model content type (The [gist](https://gist.github.com/magopian/7543724) mentioned in https://code.djangoproject.com/ticket/11154#comment:54)
- Make the user a superuser

# What will happen after this patch in terms of permissions?

- New permissions will be created for all the proxy models, using their own content type instead of the content type of the concrete model.
- Users with the new permissions will be able to access the admin pages of proxy models with an `app_label` different from the concrete model.

# What will stay the same?

- If you give a user the permissions of a concrete model, it doesn't give them any permission on an inheriting proxy model.

# Highlighted changes for review

- [x] Cherry-pick the changes of #4681
- [x] Describe with concrete examples how this page will affect permission creation
- [x] Update existing docs around proxy models and getting permissions by content type
- [x] Add the release notes
- [x] Write documentation to explain how the permissions are created for a model/proxy model
    - [x] Explain that permissions for a concrete model don't give permissions for a proxy model

carltongibson commented:
Hi @arthurio. Could you possibly squash this down to one commit (or a few commits if you think that'd help review)?

carltongibson commented on django/contrib/contenttypes/models.py:
```diff
@@ -31,6 +31,12 @@ def _get_from_cache(self, opts):
         key = (opts.app_label, opts.model_name)
         return self._cache[self.db][key]
 
+    def get_for_permission_by_model(self, model):
```
This name doesn't read right.

arthurio replied:
I agree, I initially wanted to use `Permission.objects.get_content_type_for_model` instead, but I ran into issues because the imported `ContentType` in `django.contrib.auth.models` is different than the one in [`django/contrib/auth/management/__init__.py`](https://github.com/django/django/pull/10381/files#diff-23ce0f3ba59117c95d173175f52d60c7R47):

```python
app_config = apps.get_app_config(app_label)
ContentType = apps.get_model('contenttypes', 'ContentType')
```

If someone can help me figure this out then I think it would be a much better solution to **a.** Have a better name, and **b.** Living on the `Permission` model.

I'm open to other suggestions too of course.

arthurio replied:
@auvipy It makes me think that you are getting a `permission_type` which is unclear what that could be... If there is no good name for it, I can also get rid of it and call `get_for_model(Model, for_concrete_model=False)` explicitly. It's not something that should be used a lot and I can add an entry in the docs for proxy models, what do you think?

carltongibson replied:
> I can also get rid of it and call get_for_model(Model, for_concrete_model=False) explicitly.

Yes. I think this. 

Outside of the tests, there's just a single call-site. I don't see that merits a _convenience wrapper_, rather than just passing the `kwarg`.

arthurio commented:
>Hi @arthurio. Thanks for the input here. First glance looks good. (I'll get the fine-tooth comb out when you've tidied it up. 🙂)

@carltongibson Thank you so much for taking the time to review my work! I'm done with tidying things up if you want to do a second pass.

>Not currently seeing where the permissions are created. (Your description mentions a migration to come, so...)

The permissions for proxy models are created along with the permissions for concrete models in `django/contrib/auth/management/__init__.py`. The proxy model permission are already created, but the difference this patch makes is that it will use the content type of the proxy model instead of the one of the concrete model.

```python
# myapp/models.py
class ProxyUser(get_user_model()):
    class Meta:
        proxy = True

# Before the patch
>>> Permission.objects.get(codename='add_proxyuser').content_type.model_class()
'django.contrib.auth.models.User'

# After the patch
>>> Permission.objects.get(codename='add_proxyuser').content_type.model_class()
'myapp.models.ProxyUser'
```

>Thought: The "just create the permissions" workaround doesn't seem that onerous. Perhaps documenting, and giving folks the tools to create the permissions, would be better than automatically opting everyone in? (I may have a million proxy models for which I need permissions for only a handful...)

It seems odd that you don't have the default permissions automatically created for your proxy models, unless you explicitly set `default_permissions = []` in your proxy model, they should already be created, could you double check by doing a lookup like `Permission.objects.filter(codename__endswith='yourmodel')`?

auvipy commented on django/contrib/contenttypes/models.py:
```diff
@@ -31,6 +31,12 @@ def _get_from_cache(self, opts):
+    def get_for_permission_by_model(self, model):
+        """
+        Return the `ContentType` instance for a given model to lookup a `Permission`.
+        """
```
get_for_model also doesn't sound that great IMHO....

arthurio commented on django/contrib/auth/migrations/0010_update_proxy_permissions.py:
```diff
@@ -0,0 +1,57 @@
+
+        codenames = default_proxy_codenames + custom_codenames
+
+        if opts.app_label == opts.concrete_model._meta.app_label:
```
This is the safe part, where we move the existing permissions to the new content type.

arthurio commented on django/contrib/auth/migrations/0010_update_proxy_permissions.py:
```diff
@@ -0,0 +1,57 @@
+                content_type=concrete_content_type,
+            ).update(content_type=proxy_content_type)
+
+        else:
```
This is the part where we cleanup the permissions where the `app_label` of the concrete model and the proxy model are different. Users with the old proxy permissions would now have access to admin pages they couldn't see before (but should have...). So, my idea is to just delete the old permissions and add a message to the release notes to let people know that they need to re-issue the permissions for any proxy model having a different `app_label` than their concrete model.
Other approaches:
- Do nothing: The old permissions will be listed along with the new ones and it might be confusing.
- Update the content type the same way we do it for concrete/proxy models with similar `app_label`: It might be surprising for users with permissions to all the sudden have access to pages they couldn't see before. We could also communicate that fact in the release notes...

carltongibson commented:
> It seems odd that you don't have the default permissions automatically created for your proxy models...

No, you're right. This query is just me being confused. (Better now. Thanks. 🙂)

carltongibson commented on django/contrib/admin/options.py:
```diff
@@ -468,7 +468,7 @@ def has_add_permission(self, request):
         opts = self.opts
         codename = get_permission_codename('add', opts)
-        return request.user.has_perm("%s.%s" % (opts.app_label, codename))
+        return request.user.has_perm('%s.%s' % (opts.app_label, codename))
```
Can you pull these cosmetic changes into a separate commit please? 
(I know they go with the _Patch style_ guides, but they're not related to the issue here.)

arthurio replied:
Should I put the changes in a separate commit or do them in a separate PR?

carltongibson commented on django/contrib/admin/options.py:
```diff
@@ -2138,17 +2138,20 @@ def has_delete_permission(self, request, obj=None):
-        if self.opts.auto_created:
-            opts = self.opts
+        opts = self.opts
+        if opts.auto_created:
```
Again, this probably counts as unrelated.

(Not sure it's cleaner either: `opts` is now in scope outside the `if` block...)

carltongibson commented on django/contrib/admin/options.py:
```diff
@@ -2138,17 +2138,20 @@ def has_delete_permission(self, request, obj=None):
-                request.user.has_perm('%s.%s' % (opts.app_label, get_permission_codename('view', opts))) or
-                request.user.has_perm('%s.%s' % (opts.app_label, get_permission_codename('change', opts)))
+                request.user.has_perm('%s.%s' % (opts.app_label, codename_view)) or
+                request.user.has_perm('%s.%s' % (opts.app_label, codename_change))
```
Both these lines come in at under 120 chars, so there's no real need for the temporary variables.

arthurio commented:
>OK. Looking good. I've added a few comments down to "Get rid of get_permission_for_model()". If you could address those, and adjust the docs and tests, I'll go over again.

I'm going to do a commit with those changes and will rebase once you have approved them.

>It looks like all (?) the changes to admin/options.py are merely cosmetic. I think they need to be removed from the patch.

Should I put the changes in a separate commit or do them in a separate PR?

>Thanks for the effort! Super.

You are welcome, thanks for reviewing and helping me through my first contribution.

arthurio commented:
@carltongibson I think it's ready for another pass whenever you have chance. I'd love your feedback on the docs. Also, I'll try to take a look at how the migrations are tested but would love some guidance regarding testing https://github.com/django/django/pull/10381/files#diff-72d0747292e07221fa513b781de95d5b.

arthurio commented:
@carltongibson I have removed the cosmetic changes made to `admin/options.py` and added tests for my migration. Once you are 👌 with the patch, I'll rebase on master. Also, once the PR looks good, should I squash everything in one commit or try to keep the ones made per your feedback separate?

carltongibson commented:
Hey @arthurio. Thanks for the effort. I'll give it another pass now.

carltongibson commented on docs/topics/db/models.txt:
```diff
@@ -1417,6 +1536,76 @@ different database tables).
+    ...     Student,
+    ...     for_concrete_model=False,
+    ... )
+    >>> Permission.objects.filter(content_type=content_type).values_list(
```
Missing parenthesis `)`. 

Do you want/need `values_list()` here?

carltongibson commented on docs/topics/db/models.txt:
```diff
@@ -1417,6 +1536,76 @@ different database tables).
+    True
+
+ContentType
+-----------
```
This section isn't quite right yet. 

The points you're making: 

* Concrete and proxy models compare equal.
* Content type for concrete model is fetched by default. 
* `for_concrete_model=False` gives you the distinct content type. 
* Permissions are tied to the content type. 

Each perhaps needs a sentence.

arthurio replied:
I tried to make the examples self explanatory but even if this is clear to me I can see how someone landing on the page might need more info to make sense out of it.

carltongibson commented on tests/admin_views/tests.py:
```diff
@@ -1361,10 +1362,12 @@ def test_pwd_change_custom_template(self):
-    """Return the permission object, for the Model"""
-    ct = ContentType.objects.get_for_model(Model)
-    return Permission.objects.get(content_type=ct, codename=perm)
+def get_perm(action, Model):
```
Why this refactoring? (Touches a lot of the other tests...)

arthurio replied:
The idea was to centralize the logic used to build the codename and do the lookup in this function. Instead, to reduce the diff, I'll simply update `ct = ContentType.objects.get_for_model(Model)` to use `for_concrete_model=False` in case someone adds a test with a proxy model.

carltongibson commented on tests/auth_tests/test_management.py:
```diff
@@ -976,9 +976,9 @@ def test_create_permissions_checks_contenttypes_created(self):
         `post_migrate` handler ordering isn't guaranteed. Simulate a case
         where create_permissions() is called before create_contenttypes().
         """
-        # Warm the manager cache.
```
Leave the `.`
(Makes the diff smaller if nothing else. But it's probably OK for it to be a fully-formed sentence. 🙂)

arthurio replied:
Oh it's an oversight, good catch.

carltongibson commented on tests/proxy_models/models.py:
```diff
@@ -4,6 +4,7 @@
 than using a new table of their own. This allows them to act as simple proxies,
 providing a modified interface to the data from the base class.
 """
+from django.contrib.auth.models import User as AuthUser
```
Why do we need the alias here?

arthurio replied:
There are already a `class User` and `class UserProxy` in the models...

arthurio commented on tests/proxy_models/tests.py:
```diff
@@ -188,6 +189,17 @@ def test_permissions_created(self):
+        try:
+            Permission.objects.get(content_type__model=opts.model_name,
+                                   content_type__app_label=opts.app_label,
+                                   codename=codename)
```
I'll update this line formatting.

arthurio commented:
@carltongibson After giving it some more thoughts, I decided to remove the migration file and to add the following text to the release notes:
"To keep existing group and user permissions intact, you will have to re-issue the new proxy permissions."

I'd love some feedback on the wording and also if you think it's useful or not.

And here is the initial migration I wrote for posterity (Someone could add it to their migrations **prior** to do the first migrate following the update to `Django==2.1.2`):

```python
from django.contrib.auth import get_permission_codename
from django.db import migrations


def update_proxy_model_permissions(apps, schema_editor):
    Permission = apps.get_model('auth', 'Permission')
    ContentType = apps.get_model('contenttypes', 'ContentType')

    for Model in apps.get_models():
        opts = Model._meta

        if not opts.proxy:
            continue

        concrete_content_type = ContentType.objects.get_for_model(
            Model,
            for_concrete_model=True,
        )

        proxy_default_codenames = [
            get_permission_codename(action, opts)
            for action in opts.default_permissions
        ]
        proxy_custom_codenames = {codename for codename, __ in opts.permissions}

        concrete_custom_codenames = {
            codename for codename, __ in opts.concrete_model._meta.permissions
        }

        custom_codenames = list(concrete_custom_codenames - proxy_custom_codenames)

        codenames = proxy_default_codenames + custom_codenames

        permission_queryset = Permission.objects.filter(
            codename__in=codenames,
            content_type=concrete_content_type,
        )

        if opts.app_label == opts.concrete_model._meta.app_label:
            proxy_content_type = ContentType.objects.get_for_model(
                Model,
                for_concrete_model=False,
            )

            permission_queryset.update(content_type=proxy_content_type)
        else:
            permission_queryset.delete()

class Migration(migrations.Migration):

    dependencies = [
        ('auth', '0009_alter_user_last_name_max_length'),
        ('contenttypes', '0002_remove_content_type_name'),
    ]

    operations = [
        migrations.RunPython(update_proxy_model_permissions),
    ]
```

timgraham commented on tests/proxy_models/models.py:
```diff
@@ -205,3 +206,12 @@ class Improvement(Issue):
         proxy = True
+
+
+class AuthUserProxy(AuthUser):
```
This is testing a change in `contrib.auth` so I think the test would be better in `auth_tests`.

carltongibson replied:
Does it work if you explicitly set the `app_label`?

```
class Bar(models.Model):
    name = models.CharField(max_length=255)

    class Meta:
        app_label = 'a_different_app'
```

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -2272,6 +2273,103 @@ def test_post_save_message_no_forbidden_links_visible(self):
+        },
+    }],
+)
+class AdminViewProxyModelPermissionsTest(TestCase):
```
It's not obvious to me why these tests are needed. Is there a particular concern about a bug in the admin such that this wouldn't work?

arthurio replied:
@carltongibson TL;DR; Yes this is expected.

We should see 4 errors indeed but I think it's because it fails on `setUpTestData`, therefor none of the tests of the test case are run:
```
$ ./runtests.py admin_views.tests.AdminViewProxyModelPermissionsTest

======================================================================
ERROR: setUpClass (admin_views.tests.AdminViewProxyModelPermissionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arthur/code/django/django/test/testcases.py", line 1036, in setUpClass
    cls.setUpTestData()
  File "/Users/arthur/code/django/tests/admin_views/tests.py", line 2417, in setUpTestData
    cls.viewuser.user_permissions.add(get_perm(UserProxy, get_permission_codename('view', opts)))
  File "/Users/arthur/code/django/tests/admin_views/tests.py", line 1369, in get_perm
    return Permission.objects.get(content_type=ct, codename=codename)
  File "/Users/arthur/code/django/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/arthur/code/django/django/db/models/query.py", line 408, in get
    self.model._meta.object_name
django.contrib.auth.models.Permission.DoesNotExist: Permission matching query does not exist.
```

If you revert both changes (`get_perm` and `contrib/auth/management/__init__.py`), then `get_perm` behaves the same as in `contrib/auth/management/__init__.py` so the `setUpTestData` doesn't raise an error and we get 4 failures instead of 1 error.

timgraham commented on tests/proxy_models/tests.py:
```diff
@@ -188,6 +189,19 @@ def test_permissions_created(self):
+        from django.contrib.auth.models import Permission
+        opts = AuthUserProxy._meta
+        codename = get_permission_codename('add', opts)
+        try:
```
try/except/fail is generally considered an antipattern, e.g. 7fec264e46d2a757f06f857e513072d72686cf9d. Maybe it's okay here.

timgraham commented on docs/topics/db/models.txt:
```diff
@@ -1417,6 +1417,79 @@ different database tables).
 any model field in any ancestor model.
 
+
+.. _proxy-models-permissions-topic:
```
I think this documentation might be more appropriate in topics/auth/default.html under the discussion of permission.

timgraham commented on docs/topics/db/models.txt:
```diff
@@ -1417,6 +1417,79 @@ different database tables).
+ContentType
+-----------
+
+Proxy models have their own content type which are used to create permissions
```
This discussion seems unrelated to the fix at hand.

carltongibson replied:
This is kind-of half-ref to [#22270](https://code.djangoproject.com/ticket/22270): permissions go via content types... If it's a worthwhile clarification it could go in by itself (in another PR).

arthurio replied:
@carltongibson Do you want me to move this piece to #10454 ? Or in a new PR altogether?

timgraham commented:
Could you first work on the release notes (docs/releases/2.2.txt) that explain the current migration path that you envision? It seems you changed the approach (e.g. ditched the migration) since your writing in the pull request description). I started testing this out and am confused about the extra permissions that are created. It seems my admin page still works with the old permissions in place but as a user, I'm not sure how to proceed. Thanks.

arthurio commented:
@timgraham Sorry for the confusion, I went back and forth on how to deal with the new generated permissions... To recap, I think we can take 3 approaches:

### 1. Automate with migration (update)

Update all existing proxy model permissions to use the proxy model content type.

**Pros**
- No need to deal with re-issuing permissions and deleting old ones.

**Cons**
- Users having permissions for proxy models with a different `app_label` than their concrete model will now have access to admin pages they couldn't see before.
- Need to update all calls to methods using the natural key to lookup permissions (such as `user.has_perm('app_label.view_proxymodel`).

### 2. Automate with migrations (update + delete)

Update all existing proxy model permissions to use the proxy model content type, **but** delete the ones for proxy models having a different `app_label` than their concrete model.

**Pros**
-  Users having permissions for proxy models with a different `app_label` than their concrete model will not gain access to admin pages they couldn't see before.

**Cons**
- Need to re-issue new permissions for proxy models with different `app_label` than concrete model.
- Can lead to data loss.

### 3. Don't do anything

**Pros**
- No need to update calls to methods using the natural key to lookup permissions.
- Existing permissions keep working the same way. (Can be a "con" obviously...)

**Cons**
- Need to manually deal with re-issuing new permissions and deleting old ones.

## Conclusion

`3.` is definitely the safest approach but requires (a lot of) manual work and can lead to confusion since you'll end up with two almost identical proxy model permissions.

`2.` is a mix of safe + less manual work but do require a `delete` which can cause data loss.

`1.` is the best approach in terms of permission data integrity but can lead to the surprising side effect that someone will now have access to more admin views if he had permissions for it. But it's not really a "con" to me because why would that person have the permission if the original intent wasn't to give them access to the admin page?

I personally prefer solution `1.` and I can easily add a migration with some tests if you agree. If we pick that solution, I don't think we need to update the release notes, unless you want to mention that side effect?

carltongibson commented:
Hi @arthurio. Yes, erm... sorry for the delay: constant flow of incoming... 

> Could you first work on the release notes (docs/releases/2.2.txt) 

This. Your comment is fine, but it does need to be part of the PR. You still have changes to 2.1.2.txt for example. 

(My last take was "more or less yes" but just "how do we introduce this".

I will come back to this next week. Thanks for the effort! ([v.2.2 feature freeze is not till January](https://code.djangoproject.com/wiki/Version2.2Roadmap).)

carltongibson commented:
Hi @arthurio. I've gone over again, resolving as many conversations as possible from the early history. 

Current status: 

* Looks good but...
* There's this migration question. 

First of all can you please go through each of the points Tim made and adjust the code appropriately so we can resolve those too? (Leave any where there's a genuine question.) 

Then: 

Whatever we do here the release notes are going to need a substantial entry. (We have to spell out exactly what the approach is.) 

Can you write up a guide for your favoured option 1? What do we provide? (i.e. the migration) What further steps can I take manually? (i.e. delete old permissions) What gotchas must I look out for? (i.e. unwanted access to admin pages.) Imagine you have an large existing project and need to step through it. 

From there, I would then ask on django-developers for further review, to make sure we didn't miss anything. 

For me, this solves the problem on a 9 year old ticket, so it's worth having. The question is, can we release it in a way that solves the migration and security concerns in an acceptable manner?

If the answer to that is "No" then (IMO) we need to document the method of scripting or manually creating the required permissions proxy model with the correct app label and call it day. That would obviously not as good. 

So plan: 

1. Address Tim's feedback. 
2. Write release notes, explaining update path. 
3. Check with others that it's good to go. 

Thank you  for your effort here. I know it's a lot of work (especially for a first contribution). The old tickets are always going to be the hardest. 🙂

(Let me know if you need assistance.)

arthurio commented:
@carltongibson Thank you so much for taking the time to give me your feedback, it's exactly what I needed to move forward. I have pushed 4 commits in response to feedback by you or @timgraham but there are 2 outstanding ones:
1. https://github.com/django/django/pull/10381#discussion_r229518308 Do you of a way to have models with different `app_label`s in the same test app? That would remove the confusion about why I use `auth.User` as parent model. Or maybe I should just rename the test proxy model?
2. https://github.com/django/django/pull/10381#discussion_r229520762 This is about the relevance of the test I have added for the admin page of a proxy model with a different `app_label` than its concrete model. I believe this is what's called a "regression" test and I think it is useful, but please, correct me if I'm wrong.

For the things to review in the new commits, I think the critical piece is the migration file and the tests that go with it (all in https://github.com/django/django/pull/10381/commits/1682586eeda6556d4ef318b5151395007a05a9ef).

I will start writing more detailed release notes and the draft for the email to the dev mailing list, but I will be travelling for the next few days so there might be a bit of delay.

arthurio commented:
@carltongibson Please, let me know if the changes look good and I will squash my commits. I also sent an email asking for feedback: https://groups.google.com/forum/#!topic/django-developers/6C-LZaek3nE

arthurio commented:
@carltongibson I added the backward migration as suggested by @aaugustin. Please, let me know if I can squash my commits or if you want to review the changes first.

carltongibson commented:
Hey @arthurio Please squash and rebase. I will take another (hopefully ≈last 🙂) look at this today. Super effort!

arthurio commented:
> Hey @arthurio Please squash and rebase. I will take another (hopefully ≈last 🙂) look at this today. Super effort!

Thanks @carltongibson, done!

carltongibson commented:
Hi @arthurio. I added a few edits to the release notes. Can you review? 
https://github.com/django/django/pull/10381/commits/caf0a02810052d9e95ccbe048a24f0f6f3982b77

carltongibson commented:
@arthurio: Sorry scrub the `app_label` thought: Just moving `test_permission_with_correct_content_type_created()` (with the model) into the auth_tests app works as expected.

arthurio commented:
@carltongibson Again, I used `AuthUser` because this is a model most often available. I could have picked any other model to test this behavior. The reason I chose it is because I couldn't make it work with an explicit `app_label = app_b` like you mentioned. One thing I may not have tried is to create a sub folder in the `proxy_models` test app and create my proxy model there, by adding both `proxy_models` and `proxy_models.different_app_label` to the `installed_apps`. Do you want me to try that?

carltongibson commented:
Hey @arthurio. The point is just to move the changes from the `tests/proxy_models` folder (app) into the `tests/auth_tests` folder (app). Nothing else needs to change.

arthurio commented:
>Hey @arthurio. The point is just to move the changes from the tests/proxy_models folder (app) into the tests/auth_tests folder (app). Nothing else needs to change.

Ohhh I misread, yes, it's a permission test, not a proxy model test, sounds good to me!

arthurio commented on tests/auth_tests/models/__init__.py:
```diff
@@ -15,5 +16,6 @@
     'MinimalUser', 'NoPasswordUser', 'UUIDUser', 'CustomUserNonUniqueUsername',
-    'IntegerUsernameUser', 'UserWithDisabledLastLoginField',
+    'IntegerUsernameUser', 'UserWithDisabledLastLoginField', 'Proxy',
+    'UserProxy',
```
Should I re-order the `__all__`? I'm not sure what is the logical order used currently...

arthurio replied:
I did... let met know if I shouldn't have :)

arthurio commented:
@carltongibson I moved the tests to `auth_tests` but it's really unclear to me why I'm getting those 2 failures consistently: https://djangoci.com/job/pull-requests-bionic/database=mysql,label=bionic-pr,python=python3.5/3283/ I haven't been able to reproduce locally so far with sqlite but I'll try with mysql tomorrow... Do you have any idea about what could be causing this behavior?

carltongibson commented:
Hi @arthurio. The failures are a little strange. They came up with Postgres, but I couldn't reproduce that locally, so I've squashed and rebased again. We'll see what happens...

timgraham commented:
I can reproduce on one of my machines by not the other (both Ubuntu 18.04). It might have to do with the unpredictable order in which tests are executed.

timgraham commented:
You probably need to add `ContentType.objects.clear_cache()` in `setUp()` and/or `tearDown()` of the affected classes because of `ContentType.objects.get_for_model()` caching.  See other `contenttypes_tests` for examples.

arthurio commented:
Thanks for the suggestions @timgraham and @carltongibson. It's still failing though so I'm going to try to put some logging in a separate commit to figure out what's going on.

timgraham commented on docs/releases/2.2.txt:
```diff
@@ -57,6 +57,18 @@ Minor features
+  proxy models having the same ``app_label`` as their concrete model.)
+
+  A reversible migration is provided to automatically update existing
+  permissions when running :djadmin:`migrate`.  Reviewing admin permissions for
```
Reviewing the permissions in what way?

arthurio replied:
This is related to:
>This means that users granted permissions for proxy models having a different ``app_label`` than their concrete models will now be able to correctly access corresponding admin pages.

We want to make sure they audit the list of users/groups having such permissions prior to running the migration to avoid surprises.

timgraham replied:
I think the documentation should describe the audit process in more detail. Perhaps an example would be helpful.

arthurio replied:
Something like the following?
```python
from django.contrib.contenttypes import ContentType
from django.db.models import Q

from myapp.models import MyProxyModel


opts = MyProxyModel._meta

default_permissions_codenames = [
    '%s_%s' % (action, opts.model_name)
    for action in opts.default_permissions
]

permissions_query = Q(codename__in=proxy_default_permissions_codenames)
for codename, name in opts.permissions:
    permissions_query = permissions_query | Q(codename=codename, name=name)

content_type = ContentType.objects.get_for_model(MyProxyModel)

permission_queryset = Permission.objects.filter(
    permissions_query,
    content_type=concrete_content_type,
)

for permission in permission_queryset:
    print("Permission: %s" % permission.codename)

    print("Groups")
    print(permission.group_set.all())

    print("Users")
    print(permission.user_set.all())
```

timgraham replied:
Not necessarily code. I meant describing how someone would audit and what sort of change you might make as a result.

arthurio replied:
I think we can re-order a bit the existing text to make it more clear:

```
Permissions for proxy models
----------------------------

Permissions for proxy models are now created using the content type of the
proxy model rather than the content type of the  concrete model.

A migration is provided to update existing permissions when running
:djadmin:`migrate`.

In this respect, the change is transparent for proxy models having the same
``app_label`` as their concrete model.

However, users granted permissions for proxy models having a different
``app_label`` than their concrete models will now be able to correctly
access corresponding admin pages.  Therefore, auditing `groups` and
`users` having admin permissions (e.g. For `ProxyModel`:
`add_proxymodel`, `view_proxymodel`, `change_proxymodel`,
`delete_proxymodel`) for such proxy models prior to running
:djadmin:`migrate` is recommended. You may remove the permission
from the list of group or user permissions to prevent any unforeseen
access leak and carefully re-assign them once the :djadmin:`migrate`
command has been run.
```

Do you think that'd be enough or do you want more details about **how** to audit?

arthurio replied:
🤔 It's not really an "access leak" but more a conscious choice of "fixing" the permissions and letting users get access to the relevant admin pages **automatically**. Let me try to rephrase one more time...

```
Permissions for proxy models
----------------------------

Permissions for proxy models are now created using the content type of the
proxy model rather than the content type of the  concrete model.

A migration is provided to update existing permissions when running
:djadmin:`migrate`.

In this respect, the change is transparent for proxy models having the same
``app_label`` as their concrete model.

However, users granted permissions for proxy models having a different
``app_label`` than their concrete models will now be able to correctly
access corresponding admin pages.  Auditing permissions of groups and
users for such proxy models (``[add|view|change|delete]_myproxy``)
prior to running :djadmin:`migrate` is recommended as you might want to
notify affected users of the changes to their admin homepage beforehand.
If you want to automatically give access to admin pages of such proxy
models, you can remove their permissions from groups and users, and
re-assign them after having run :djadmin:`migrate`.
```

arthurio replied:
It's a typo, I meant "If you **don't** want to ...".

timgraham replied:
Thanks, that was my assumption. I removed that sentence since it seems obvious.

arthurio commented on docs/releases/2.2.txt:
```diff
@@ -55,7 +55,14 @@ Minor features
+  their own content type. Existing permissions will be updated automatically
+  when running :djadmin:`migrate`. Users granted permissions for proxy models
+  having a different ``app_label`` than their concrete models will now be able
+  to access their admin pages accordingly. Reviewing permissions for such proxy
```
Potentially remove the permissions from users/groups if there is a security concern prior to migrate.

arthurio commented:
>- Admin view tests are lacking as noted in the comments I put in the tests.

I see you are making edits, do you want me to update them now?

>- Release notes should discuss changes to permissions checking (User.has_perm()) that the tests demonstrate?

Yes we could add something along those lines:

```
For `User.has_perm` and `User.has_perms`, make sure that the permission
string is using the ``app_label`` of the proxy model.
```

timgraham commented:
Yes, you can edit this now. The release note should be more along the lines of "update the permission string from x to y."

carltongibson commented:
Super! Well done @arthurio on a great first contribution! (Top marks for extended effort.) Thanks for guiding it over the line Tim!

arthurio commented:
Thank you very much @carltongibson and @timgraham for shepherding my work, looking forward to more contributions!
